src/services/mapCode.ts

```diff
@@ -0,0 +1,365 @@
+    }
+}
+
+function placeClassNodeGroup(file: SourceFile, changeTracker: ChangeTracker, nodes: NodeArray<ClassElement> | NodeArray<TypeElement>, focusLocations?: TextSpan[][]) {
+    // We have one or more class-ish members.
+    //
+    // 1. If focusLocations is null/undefined, find the first class in the
+    //    SourceFile and insert/replace into that class.
+    // 2. If we have focusLocations, go over each one and:
+    //   a. Find the nearest class or interface declaration that contains the focusLocation.
+    let classOrInterface: ClassLikeDeclaration | InterfaceDeclaration | undefined;
+    if (!focusLocations || !focusLocations.length) {
+        for (const stmt of file.statements) {
+            if (isClassLike(stmt)) {
+                classOrInterface = stmt;
+                break;
+            }
+            else if (stmt.kind === SyntaxKind.InterfaceDeclaration) {
+                classOrInterface = stmt as InterfaceDeclaration;
+                break;
+            }
+        }
+    }
+    else {
+        let node: Node | undefined;
+        top: for (const locationGroup of focusLocations) {
+            for (const location of locationGroup) {
+                node = findAncestor(getNodeAtPosition(file, location.start), node => {
+                    return isClassLike(node) || node.kind === SyntaxKind.InterfaceDeclaration;
+                });
+                if (node && node.kind === SyntaxKind.InterfaceDeclaration) {
+                    classOrInterface = node as InterfaceDeclaration;
+                    break top;
+                }
+                else if (node) {
+                    classOrInterface = node as ClassLikeDeclaration;
+                    break top;
+                }
+            }
+        }
+    }
+
+    if (!classOrInterface) {
+        throw new Error("Failed to find a class or interface to map the given code into.");
+    }
+    else {
+        const classEls: ClassElement[] = [];
+        const typeEls: TypeElement[] = [];
+        top: for (const node of nodes) {
+            for (const member of classOrInterface.members) {
+                if (matchNode(node, member)) {
+                    // If we have corresponding nodes, replace the old one with the new member.
+                    changeTracker.replaceNode(file, member, node);
+                    continue top;
+                }
+            }
+            // Otherwise, we insert this node at the end of the class/interface.
+            if (isClassElement(node)) {
+                classEls.push(node);
+            }
+            else if (isTypeElement(node)) {
+                typeEls.push(node);
+            }
+        }
+        if (classEls.length || typeEls.length) {
+            if (classOrInterface.members.length === 0) {
+                if (classEls.length) {
+                    const newNode = { ...classOrInterface, members: factory.createNodeArray(classEls) };

```

Comment:
what if `typeEls.length > 0` here? Is that impossible because type members and class members can't occur in the same class?

Reply:
we simply choose to prioritize classEls.